docs/reference/modules/discovery/bootstrapping.asciidoc

```diff
@@ -10,11 +10,17 @@ data folder and freshly-started nodes that are joining an existing cluster
 obtain this information from the cluster's elected master. 
 
 The initial set of master-eligible nodes is defined in the
-<<initial_master_nodes,`cluster.initial_master_nodes` setting>>. This is a list
-of the <<node.name,node names>> or IP addresses of the master-eligible nodes in
-the new cluster. If you do not configure `node.name` then it is set to the
-node's hostname, so in this case you can use hostnames in
-`cluster.initial_master_nodes` too.
+<<initial_master_nodes,`cluster.initial_master_nodes` setting>>. This should be
+set to a list containing one of the following items for each master-eligible
+node:
+
+- The <<node.name,node name>> of the node.
+- The node's hostname if `node.name` is not set, because `node.name` defaults
+  to the node's hostname.
+- The IP address of the node's <<modules-transport,publish address>> which is
+  normally the IP address to which <<common-network-settings,`network.host`>>
+  resolves but <<advanced-network-settings,this can be overridden>>.
+- The IP address and port of the node's publish address, in the form `IP:PORT`.
```

Comment:
sorry, I still don't like this. Instead I suggest something along the lines of:

```
- The <<node.name,node name>> of the node, which defaults to the node's hostname if
  `node.name` is not set.
- The IP address of the node's <<modules-transport,publish address>>, which is the IP
  address that this node announces itself to other nodes and which the other nodes use to
  communicate with this node. In case where multiple nodes are running on the same host and
  using the same publish address, this can be further qualified with the port in the form
  `IP:PORT`. Setting unique node names and configuring cluster bootstrapping via the
  node names should however be preferred.
```

Reply:
Are you sure we want to express this kind of preference for `node.name` over the IP address? Why do we have this preference? It seems like that might push some users down a path of trying harder than necessary to pin down their node names simply to avoid the non-preferred, but easier-for-them, IP addresses.

Reply:
Selecting proper node names should be a good practice (even it is just consciously not setting it and leaving it to the default, i.e., the host name). Given this great identifier of nodes, why not  use it instead of something more cryptic like IPs?

That being said, my intention was for the last sentence ("Setting unique node names and configuring cluster bootstrapping via the node names should however be preferred") to apply for the cases where you run multiple nodes on the same host ("In case where multiple nodes are running on the same host and using the same publish address..."). In that case, it's truly important to use unique node names to avoid confusion.

Reply:
Ok, I added qualifiers of the form `if it is not possible to use the `node.name` of the node`.

I would like to keep the four options as four separate things, rather than combining them down to two as you suggested. My thinking is that users should generally pick one of these four strategies consistently across their cluster; if we only give two options with embedded conditionals then I think it's harder to see the full choice that needs to be made.